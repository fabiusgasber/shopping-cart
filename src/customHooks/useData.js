import { useEffect, useState } from "react";

const useData = (url) => {
  const [data, setData] = useState(null);
  const [error, setError] = useState(null);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    fetch(url)
      .then((res) => {
        if (!res.ok) {
          throw new Error("server error");
        }
        return res.json();
      })
      .then((json) => setData(json))
      .catch((err) => setError(err))
      .finally(() => setLoading(false));
  }, [url]);

  return { data, error, loading };
};

export default useData;
