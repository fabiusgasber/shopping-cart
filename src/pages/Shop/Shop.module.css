.main {
    display: flex;
    flex-direction: column;
    gap: 2rem;
}

.main h1 {
    align-self: center;
}

.productSection {
    display: grid;
    grid-template-columns: repeat(auto-fill, minmax(250px, 1fr));
    grid-auto-rows: 1fr;
    gap: 5rem;
}

@media (max-width:720px) {
    .main {
        align-items: center;
    }
}