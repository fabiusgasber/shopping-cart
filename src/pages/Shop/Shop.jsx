import ProductCard from "../../components/ProductCard/ProductCard";
import styles from "./Shop.module.css"
import useData from "../../customHooks/useData";
import ErrorPage from "../ErrorPage/ErrorPage"


const Shop = () => {

    const { data, error, loading } = useData("https://fakestoreapi.com/products?limit=12");

    if(loading) return <h1>...Loading items</h1>
    if(error) return <ErrorPage />
    return (
        <main className={styles.main} aria-live="polite">
                <h1>Our collection</h1>
                <section className={styles.productSection}>
                { data.map((item) => (
                    <ProductCard item={item} to={`/product/${item.id}`} key={item.id} url={item.image} alt="" caption={{ text: item.title, price: item.price }} />
                )) }
                </section>
        </main>
        )
    ;
}

export default Shop;