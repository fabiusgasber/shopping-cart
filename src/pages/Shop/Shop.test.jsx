import { describe, it, expect } from "vitest";
import { render, screen, renderHook, waitFor } from "@testing-library/react";
import Shop from "./Shop";
import useData from "../../customHooks/useData";
import { MemoryRouter } from "react-router-dom";

describe("Shop page", () => {

    it("renders shop", () => {
        const { container } = render(<MemoryRouter><Shop /></MemoryRouter>);
        expect(container).toMatchSnapshot();
    })

    it("renders an initial loading state", () => {
        render(<MemoryRouter><Shop /></MemoryRouter>);
        const heading = screen.getByRole("heading").textContent;
        expect(heading).toMatch("Loading");
    })

    it("renders shop items", async () => {
        render(<MemoryRouter>
            <Shop />
            </MemoryRouter>);
       await waitFor(async () => {
        const figures =  await screen.findAllByRole("figure");
        expect(figures.length).toBeGreaterThan(0);
        })
    })

})

describe("useData()", () => {

    it("should return the initial values for data, error and loading when called with no url", async () => {
       const { result }  = renderHook(() => useData());
       await waitFor(() => {
       const { loading, error, data } = result.current
       expect(loading).toBeTruthy();
       expect(error).toBeNull();
       expect(data).toBeNull();
       })
    })

    it("fetches dummy json data", async () => {
        const mockResponse = {
    "userId": 1,
    "id": 1,
    "title": "delectus aut autem",
    "completed": false
    }
    const testUrl = "https://jsonplaceholder.typicode.com/todos/1";
    const { result } = renderHook(() => useData(testUrl));
    await waitFor(() => {
    expect(result.current.data).toEqual(mockResponse);
    });
    })

    it("throws error when url is not available", async () => {

        const { result } = renderHook(() => useData("https://fakeurlblabla.com"));
        await waitFor(() => {
            const { error, loading, data } = result.current;
            expect(loading).toBeFalsy();
            expect(error).not.toBeNull();
            expect(data).toBeNull();
        })
         
    })
})