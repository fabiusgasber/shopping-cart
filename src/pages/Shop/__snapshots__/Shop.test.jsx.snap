// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Shop page > renders shop 1`] = `
<div>
  <h1>
    ...Loading items
  </h1>
</div>
`;
